Tasks/Common/azure-arm-rest/azure-arm-endpoint.ts

```diff
@@ -0,0 +1,111 @@
+import tl = require('vsts-task-lib/task');
+import Q = require('q');
+import webClient = require("./webClient");
+import { AzureEndpoint } from "./azureModels";
+
+export class AzureRMEndpoint {
+    public endpoint: AzureEndpoint;
+    private _connectedServiceName: string;
+    // Add an entry here and separate function for each new environment
+    private _environments = {
+        'AzureStack': 'azurestack'
+    }
+
+    constructor(connectedServiceName: string) {
+        this._connectedServiceName = connectedServiceName;
+        this.endpoint = null;
+    }
+
+    public async getEndpoint() {
+        let dataDeferred = Q.defer<AzureEndpoint>();
+        if(!!this.endpoint) {
+            return this.endpoint;
+        }
+        else {
+            this.endpoint = {
+                subscriptionID: tl.getEndpointDataParameter(this._connectedServiceName, 'subscriptionid', true),
+                subscriptionName: tl.getEndpointDataParameter(this._connectedServiceName, 'subscriptionname', true),
+                servicePrincipalClientID: tl.getEndpointAuthorizationParameter(this._connectedServiceName, 'serviceprincipalid', false),
+                servicePrincipalKey: tl.getEndpointAuthorizationParameter(this._connectedServiceName, 'serviceprincipalkey', false),
+                environmentAuthorityUrl: tl.getEndpointDataParameter(this._connectedServiceName, 'environmentAuthorityUrl', true),
+                tenantID: tl.getEndpointAuthorizationParameter(this._connectedServiceName, 'tenantid', false),
+                url: tl.getEndpointUrl(this._connectedServiceName, true),
+                environment: tl.getEndpointDataParameter(this._connectedServiceName, 'environment', true),
+                activeDirectoryResourceID: tl.getEndpointDataParameter(this._connectedServiceName, 'activeDirectoryServiceEndpointResourceId', true)
+            } as AzureEndpoint;
+
+            if(this.endpoint.environment != null && ( !this.endpoint.environmentAuthorityUrl || !this.endpoint.activeDirectoryResourceID)) {
```

Comment:
Check for Azure Stack can be moved above instead of nested check.

Reply:
done